src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,26 +30,152 @@ public EntityMappingImpl( final Graph mappingDescription ) {
+				continue;
+			}
+			final Set<Node> locals = g2lMap.get(workingTriple.getSubject());
+			if(locals != null) {
+				// There is already at least one local URI for this global URI,
+				// We don't want to override it, so we'll add the new one to the set.
+				locals.add(workingTriple.getObject());
+			} else {
+				// g2lmap doesn't contain any local URI for this global yet
+				final Set<Node> localSet = new HashSet<>();
+				g2lMap.put( workingTriple.getSubject(), localSet );
+				localSet.add(workingTriple.getObject());
+			}
+			
+			// The same for l2g.
+			final Set<Node> globals = l2gMap.get(workingTriple.getObject());
+			if(globals != null) {
+				globals.add(workingTriple.getSubject());
+			} else {
+				final Set<Node> newGlobalSet = new HashSet<>();
+				l2gMap.put( workingTriple.getObject(), newGlobalSet );
+				newGlobalSet.add(workingTriple.getSubject());
+			}
 		}
 	}
 
 	@Override
 	public Set<TriplePattern> applyToTriplePattern( final TriplePattern tp ) {
-		// TODO: implement this function
-		return Collections.singleton(tp);
+		final Set<Triple> workingSet = new HashSet<Triple>();
+		if(tp.asJenaTriple().getSubject().isURI()) {
+			final Set<Node> localSubjects = g2lMap.get( tp.asJenaTriple().getSubject() );
+			if( localSubjects != null ) {
+				for (final Node localSubject : localSubjects) {
+					final Triple workingTriple = new Triple(localSubject,tp.asJenaTriple().getPredicate(), tp.asJenaTriple().getObject());
+					workingSet.add(workingTriple);
+				}
+			} else {
+				workingSet.add(tp.asJenaTriple());
+			}
+		} else {
+			workingSet.add(tp.asJenaTriple());
+		}
+		final Set<TriplePattern> resultSet = new HashSet<TriplePattern>();
+			if(tp.asJenaTriple().getObject().isURI()) {
+			final Set<Node> localObjects = g2lMap.get(tp.asJenaTriple().getObject());
+			if(g2lMap.containsKey(tp.asJenaTriple().getObject())) {
+				for (final Triple workingTriple : workingSet) {
+					for (final Node localObject : localObjects) {
+						final Triple resultTriple = new Triple(workingTriple.getSubject(), workingTriple.getPredicate(), localObject);
+						resultSet.add(new TriplePatternImpl(resultTriple));
+					}
+				}
+				return resultSet;
+			} else {
+				for (final Triple workingTriple : workingSet ) {
+					resultSet.add(new TriplePatternImpl(workingTriple));
+				}
+				return resultSet;
+			}
+		} else {
+			for (final Triple workingTriple : workingSet ) {
+				resultSet.add(new TriplePatternImpl(workingTriple));
+			}
+			return resultSet;
+		}
 	}
 
 	@Override
 	public Set<SolutionMapping> applyToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		final Set<Map<Var,Node>> cartesianProduct = new HashSet<>();
+		final Binding binding = sm.asJenaBinding();
+		final Iterator<Var> it = binding.vars();
+		
+		final Var firstVar = it.next(); // Create a number of single var-node combinations to populate the set with.
```

Comment:
There is potential for a bug here. This code assumes that the given solution mapping `sm` is defined for at least one variable. However, `sm` may also be the empty solution mapping. This case needs to be handled:
```suggestion
		final Binding binding = sm.asJenaBinding();

		// If the given solution mapping is the empty mapping, simply return it unchanged.
		if ( binding.isEmpty() ) {
			return Collections.singleton(sm);
		}

		final Set<Map<Var,Node>> cartesianProduct = new HashSet<>();
		final Iterator<Var> it = binding.vars();
		
		final Var firstVar = it.next(); // Create a number of single var-node combinations to populate the set with.
```